test/AtlasErrorBubble.t.sol

```diff
@@ -0,0 +1,135 @@
+            .withGovernance(governanceEOA)
+            .withCallConfig(callConfig)
+            .buildAndIntegrate(atlasVerification);
+    }
+
+    function test_bubbles_app_revert_inside_atlas_errors_preOps() public {
+        // Require preOps and allow reuse so Atlas reverts and bubbles the revert payload
+        CallConfig memory cfg = new CallConfigBuilder()
+            .withRequirePreOps(true)
+            .withReuseUserOp(true)
+            .build();
+
+        dAppControl = _deployControl(cfg);
+        noopSolver = new NoopSolver();
+
+        // Configure DAppControl to revert during preOps
+        dAppControl.setPreOpsShouldRevert(true);
+
+        // Build a minimal metacall
+        UserOperation memory userOp = new UserOperationBuilder()
+            .withFrom(userEOA)
+            .withTo(address(atlas))
+            .withValue(0)
+            .withGas(1_000_000)
+            .withMaxFeePerGas(tx.gasprice + 1)
+            .withNonce(address(atlasVerification), userEOA)
+            .withDeadline(block.number + 2)
+            .withDapp(address(dAppControl))
+            .withControl(address(dAppControl))
+            .withCallConfig(IDAppControl(address(dAppControl)).CALL_CONFIG())
+            .withDAppGasLimit(IDAppControl(address(dAppControl)).getDAppGasLimit())
+            .withSolverGasLimit(IDAppControl(address(dAppControl)).getSolverGasLimit())
+            .withBundlerSurchargeRate(IDAppControl(address(dAppControl)).getBundlerSurchargeRate())
+            .withSessionKey(address(0))
+            .withData("")
+            .signAndBuild(address(atlasVerification), userPK);
+
+        SolverOperation[] memory solverOps = new SolverOperation[](1);
+        solverOps[0] = new SolverOperationBuilder()
+            .withFrom(solverOneEOA)
+            .withTo(address(atlas))
+            .withValue(0)
+            .withGas(1_000_000)
+            .withMaxFeePerGas(userOp.maxFeePerGas)
+            .withDeadline(userOp.deadline)
+            .withSolver(address(noopSolver))
+            .withControl(userOp.control)
+            .withUserOpHash(userOp)
+            .withBidToken(userOp)
+            .withBidAmount(0)
+            .withData("")
+            .signAndBuild(address(atlasVerification), solverOnePK);
+
+        DAppOperation memory dappOp = new DAppOperationBuilder()
+            .withFrom(governanceEOA)
+            .withTo(address(atlas))
+            .withNonce(address(atlasVerification), governanceEOA)
+            .withDeadline(userOp.deadline)
+            .withControl(userOp.control)
+            .withBundler(address(0))
+            .withUserOpHash(userOp)
+            .withCallChainHash(userOp, solverOps)
+            .signAndBuild(address(atlasVerification), governancePK);
+
+        uint256 gasLim = _gasLim(userOp, solverOps);
+
+        // Execute and capture revert data
+        vm.prank(userEOA);
+        try atlas.metacall{ gas: gasLim }(userOp, solverOps, dappOp, address(0)) returns (bool) {
```

Comment:
I think its cleaner to do

`(bool res, bytes memory data) = address(atlas).call(...)`, then checks below, than the try-catch pattern here.

The test should also revert if metacall does not revert I think.

Reply:
fixed: 0a1429dfd3c00ab115818676b0f1553a34f171c6